clients/src/main/java/org/apache/kafka/clients/consumer/PartitionAssignor.java

```diff
@@ -0,0 +1,251 @@
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * This interface is used to define custom partition assignment for use in
+ * {@link org.apache.kafka.clients.consumer.KafkaConsumer}. Members of the consumer group subscribe
+ * to the topics they are interested in and forward their subscriptions to a Kafka broker serving
+ * as the group coordinator. The coordinator selects one member to perform the group assignment and
+ * propagates the subscriptions of all members to it. Then {@link #assign(Cluster, GroupSubscription)} is called
+ * to perform the assignment and the results are forwarded back to each respective members
+ *
+ * In some cases, it is useful to forward additional metadata to the assignor in order to make
+ * assignment decisions. For this, you can override {@link #subscriptionUserdata(Set)} and provide custom
+ * userData in the returned Subscription. For example, to have a rack-aware assignor, an implementation
+ * can use this user data to forward the rackId belonging to each member.
+ */
+public interface PartitionAssignor {
+
+    /**
+     * Return serialized data that will be included in the serializable subscription object sent in the
+     * joinGroup and can be leveraged in {@link #assign(Cluster, GroupSubscription)} ((e.g. local host/rack information)
+     *
+     * @return Non-null optional join subscription user data
+     */
+    default ByteBuffer subscriptionUserdata(Set<String> topics) {
+        return ByteBuffer.wrap(new byte[0]);
+    }
+
+    /**
+     * Perform the group assignment given the member subscriptions and current cluster metadata.
+     * @param metadata Current topic/broker metadata known by consumer
+     * @param subscriptions Subscriptions from all members including metadata provided through {@link #subscriptionUserdata(Set)}
+     * @return A map from the members to their respective assignment. This should have one entry
+     *         for all members who in the input subscription map.
+     */
+    GroupAssignment assign(Cluster metadata, GroupSubscription subscriptions);
+
+    /**
+     * Callback which is invoked when a group member receives its assignment from the leader.
+     * @param assignment The local member's assignment as provided by the leader in {@link #assign(Cluster, GroupSubscription)}
+     */
+    void onAssignment(Assignment assignment);
+
+    /**
+     * Callback which is invoked when a group member receives its assignment from the leader.
+     * @param assignment The local member's assignment as provided by the leader in {@link #assign(Cluster, GroupSubscription)}
+     * @param generation The consumer group generation associated with this partition assignment (optional)
+     */
+    default void onAssignment(Assignment assignment, int generation) {
```

Comment:
I think we can actually take this out with onPartitionsLost, but leaving it in for now to avoid potentially breaking tests

Reply:
Ack, I will take it out in the other PR.

Reply:
@hachikuji @ableegoldman My understanding is that: old sticky assignor would still be on the old interface and not be migrated to the new interface; and hence for new interface there's no need to expose generation id?

Reply:
My thoughts were this: all existing assignors are migrated to the new interface. When calling onPartitionsLost, or anytime we are updating the assignment in `ConsumerCoordinator`, we also call `assignor.onAssignment(Assignment)`. So if a member using the old StickyAssignor misses a generation, its assignor state is effectively wiped and it can't claim partitions from a previous generation.